### ISSUE TITLE
Update hackage-server for Cabal 3

### ISSUE THREAD

**Issue by bgamari** [haskell/hackage-server#852]
This ticket tracks the progress of porting `hackage-server` to `Cabal-3`.

**Comment by bgamari** [comment]
I started working on this a while ago (see https://github.com/bgamari/hackage-server/tree/cabal-3) but admittedly didn't make it very far. The problem was that I wasn't particularly happy with any of the options for rewriting the `Pretty` and `Parsec` instances for `Day` and `UTCTime` in `Distribution.Server.Framework.Instances`.

@hvr has also said that he wanted someone to verify that the `Binary` instances remained compatible (although I'll admit I don't know which `Binary` instances he was referring to; I don't see any such non-trivial instances in `hackage-server`).

**Comment by bgamari** [comment]
Ahh, I see. `hackage-server` uses `cereal`'s `Serialize` class, not `Binary`. This is likely what @hvr was referring to.

**Comment by phadej** [comment]
See https://github.com/hackage-trustees/hackage-cli/pull/13

**Comment by tfausak** [comment]
I started looking at this too. You can take a look at [my branch](https://github.com/tfausak/hackage-server/tree/gh-852-cabal-three), but it's still very much a work in progress. 

I've spent about a day's worth of work on it so far. I have been having an extremely hard time making the upgrade to Cabal 3. So many things have changed, moved, or been deleted that it's hard to write code that's both backwards and forwards compatible. On top of that, some of the code that is forwards compatible is simply broken, like the `Parsec` instance for `PackageIdentifier`.

**Comment by tfausak** [comment]
I see a bunch of commits were recently pushed to master that seem to start addressing this issue: f3c5eb85d5f78330fc7d4557bcceb70ea76a91e0, 11136d332acabdec74d21315fe76e6a28660afcf, 70de659c8e3404f1523d40264b25d6f9bf5c1191, b58a7df65deb66c453862421a7d66778e70bbf86, 8924ecad1f06f1dd3d678b2224cef2c37d203225, add48b9cd24282d7b8c43593796f882bfad968be, d003d77535cbce41ad715dfabb9b660cd78a052e. Should I stop working on this?

**Comment by tfausak** [comment]
I'm not working on this anymore. As far as I can tell, [the wip/cabal-3.0 branch](https://github.com/haskell/hackage-server/compare/wip/cabal-3.0) is making good headway.

**Comment by dwijnand** [comment]
Oh.  I wasn't aware of this thread or the progress reported here...

I wanted to create such a thread to mark trailing TODOs on the progress being made in wip/cabal-3.0:
* get the code passing on CI: https://github.com/haskell/hackage-server/commit/b58d61ee698ef4f764054ce255c925194222da1b
* implement `Pretty` and `Parsec` for `Time`: https://github.com/haskell/hackage-server/commit/088bfed4d28f20637fc7cdfa143589cb334fe0e4#diff-0251fd4f91872f46f89c3c26c76750b2R241-R242
* implement the `flagAssignment` field of `BuildReport`'s `FieldGrammar`: https://github.com/haskell/hackage-server/commit/088bfed4d28f20637fc7cdfa143589cb334fe0e4#diff-0251fd4f91872f46f89c3c26c76750b2R254

**Comment by bgamari** [comment]
What remains to be done here, @dwijnand?

**Comment by dwijnand** [comment]
I'm going to sync up with @hvr on this soon, but it builds already against Cabal 3 (in the wip/cabal-3.0 branch), what's left is upgrading the logic to the changes Cabal 3 brought.

**Comment by swamp-agr** [comment]
Hello @dwijnand,

I kindly ask you, do you have any news regarding this topic?
Do you know, who is planning to work on the issue?

Thanks, 
Andrey

**Comment by dwijnand** [comment]
No, sorry, no news.  I'm not sure who (if anyone) has taken up this task.

**Comment by alexbiehl** [comment]
@dwijnand can you summarize what's still missing? From the commits on your branch I can see you managed to build hackage with Cabal 3. What's left?

**Comment by dwijnand** [comment]
If I remember correctly from my chat with @hvr there was the whole aspect of the behavioural changes in Cabal 3 to apply and verify, with the hackage server's data.  I don't recall the details and/or if there were other remaining tasks, sorry.

**Comment by alexbiehl** [comment]
Thank you for your quick response. I have a call scheduled with Herbert in the hope we can come up with a validation plan.

**Comment by tchoutri** [comment]
@alexbiehl Hi Alexander! Are there any news on this front? :)

**Comment by cartazio** [comment]
And what if anything can other contributors do to help facilitate?

**Comment by bgamari** [comment]
I would encourage anyone who works on this to proactively summarize their status as updates are made. One of the things that makes this issue quite hard to resolve is that there is generally a lack of transparency into what has been done by who, what remains to be done, and what can be punted into the future.

**Comment by recursion-ninja** [comment]
I need to upload a new version of a package to `hackage.org` which uses features in the `.cabal` file format that were introduced in `Cabal-3.0`, notably multiple public libraries.

How long should I expect to wait before I'm able to upload the new version of my package?

**Comment by alexbiehl** [comment]
Sorry for being late. Herbert and I had the call. Two things necessary:

1. Making Hackage compile with Cabal 3 is done thanks to @dwijnand. What's missing is proper validation. 

2. Hackage as running (see hackage-central branch) requires a library cabal-parsers. cabal-parsers is being used to parse and validate Cabal files. It needs to support every version of Cabal out there. I have opened a PR to support Cabal 3 [here](https://github.com/haskell-hvr/cabal-parsers/pull/1).  

@hvr Can you give us an update on where you are with the validation?

**Comment by gbaz** [comment]
So once cabal-parsers is updated, that suffices to support validation, or are we missing something else?

**Comment by gbaz** [comment]
@alexbiehl can you open a PR for that branch, or do you think that there are any further issues? i.e. should we just merge it?

**Comment by alexbiehl** [comment]
I think it makes sense to get this thing merged. However, I am unsure about deployment without proper test using current production state. The concern is that due to auto-derived instances serialization might be broken for some types. Hitting such a case would corrupt the state. 

It's not the right time for it but I will say it nevertheless: never serialize data without explicit contract. (Learned the hard way at @shlevy's school of hard knocks)

**Comment by fumieval** [comment]
Is there a recommended way to test hackage server? I built an older version (86e5e024e4369dc00cd35827fcf565b172728f74) of Hackage, started it locally with an HTTPS reverse proxy (since cabal-install can only upload through HTTPS regardless of `secure`), and registered a test user successfully. However I didn't get to upload a package; `cabal` (cabal-install 3.0.0.0 with a hack to add `--insecure`) is stuck at

```                                                     
/usr/bin/curl --insecure --config - 'https://localhost:4433/packages/candidates' --insecure --form 'package=@dist-newstyle/sdist/deriving-aeson-0.2.2.tar.gz' --write-out '
%{http_code}' --user-agent 'cabal-install/3.0.0.0 (linux; x86_64)' --silent --show-error --header 'Accept: text/plain' --location                                                       
```

It's hard to tell what's going on because hackage-server prints nothing.

**Comment by phadej** [comment]
@fumieval upload through the web interface (if you are not particularly testing cabal-install upload functionality).

**Comment by expipiplus1** [comment]
> I think it makes sense to get this thing merged. However, I am unsure about deployment without proper test using current production state. The concern is that due to auto-derived instances serialization might be broken for some types. Hitting such a case would corrupt the state.

What's the plan for getting this merged? Super looking forward to using this.

**Comment by Kleidukos** [comment]
@alexbiehl @hvr Hi, I hope you're alright in these difficult times. 
I was wondering what were the news on this front. :)

**Comment by gbaz** [comment]
Ok this is deployed now. Thanks everyone for your help!

**Comment by mitchellwrosen** [comment]
I see hackage still rejects multi-public-library packages because cabal emits a warning. Is that intentional?

```
Error: Invalid package

visibility is experimental feature (issue #5660)
```

**Comment by expipiplus1** [comment]
@mitchellwrosen perhaps this will be enabled once https://github.com/haskell/cabal/issues/5660 is closed

**Comment by philderbeast** [comment]
I tried today to upload a package candidate to hackage with multiple public libraries that uses the `package:library` format in some if its dependencies but this candidate was rejected:

```
Invalid package
plugins-for-blobs-0.1.0/plugins-for-blobs.cabal:939:25:
  colon specifier is experimental feature (issue #5660)
```

A snippet of the cabal file in that region:

```cabal
  build-depends:
      base
    , plugins-for-blobs:uom-plugin
    , plugins-for-blobs:uom-quantity
    , plugins-for-blobs:uom-th
    , tasty
    , tasty-hunit
```

**Comment by andreasabel** [comment]
Maybe open a new issue for the multi-libs problem?

**Comment by philderbeast** [comment]
@andreasabel I will open a new issue with the reproduction steps I have.

I removed the `package:library` format use and tried uploading again finding visibility is rejected:

```
Invalid package
plugins-for-blobs-0.1.0/plugins-for-blobs.cabal:444:21: visibility is experimental feature (issue #5660)
```

**Comment by recursion-ninja** [comment]
I've been waiting for over 2 years to upload an updated version of `bv-little` to hackage. Long ago I refactored the codebase to utilize the multi-library package feature and an unwilling to revert the changes or maintain two branches with different build APIs. I would **greatly** appreciate prioritizing *at least* a road map and timeline for hackage support for publishing multi-public-library packages.

**Comment by gbaz** [comment]
I believe the new cabal library should remove this warning and allow upload of such packages. Two things need to happen first. First: the new cabal library 3.8 needs an official release. Second, we need to update hackage-server to use the 3.8 library following the official cabal release, and redeploy.

Note that this was all by design, as a series of open issues on cabal meant that multi-lib packages were still experimental and not complete, and correspondingly hackage server enforced the cabal warnings. Some good work by some cabal contributors (thanks @fgaz and @grayjay !) and a fair amount of sleuthing fixed the largest outstanding issues with multi-libs. The following is the cabal commit in 3.8 which removes the warning: https://github.com/haskell/cabal/issues/6801

**Comment by recursion-ninja** [comment]
@gbaz Delighted to hear the timeline is so imminent!
